drivers/serial/uart_stm32.c

```diff
@@ -652,30 +677,44 @@ static int uart_stm32_fifo_read(const struct device *dev, uint8_t *rx_data,
 static void uart_stm32_irq_tx_enable(const struct device *dev)
 {
 	USART_TypeDef *UartInstance = UART_STRUCT(dev);
+#ifdef CONFIG_PM
 	struct uart_stm32_data *data = DEV_DATA(dev);
-
-	while (atomic_cas(&data->tx_lock, 0, 1) == false) {
-	}
+	int key;
+#endif
 
 #ifdef CONFIG_PM
+	key = irq_lock();
 	data->tx_poll_stream_on = false;
+	data->tx_int_stream_on = true;
 	uart_stm32_pm_constraint_set(dev);
 #endif
 	LL_USART_EnableIT_TC(UartInstance);
+
+#ifdef CONFIG_PM
+	irq_unlock(key);
+#endif
 }
 
 static void uart_stm32_irq_tx_disable(const struct device *dev)
 {
 	USART_TypeDef *UartInstance = UART_STRUCT(dev);
+#ifdef CONFIG_PM
 	struct uart_stm32_data *data = DEV_DATA(dev);
+	int key;
 
-	LL_USART_DisableIT_TC(UartInstance);
+	key = irq_lock();
```

Comment:
Is there a real reason to have a different behavior w/ and w/o PM ?

Reply:
Yes, these different behaviors are used to deal with pm_constraint to prevent suspension (see e3f4907efe6438a4b28f17167c58a2e25865c295)